plugins/woocommerce/src/Internal/Features/FeaturesController.php

```diff
@@ -172,22 +98,85 @@ public function __construct() {
-		$this->compatibility_info_by_plugin  = array();
-		$this->compatibility_info_by_feature = array();
-
-		$this->features = $features;
-
-		foreach ( array_keys( $this->features ) as $feature_id ) {
-			$this->compatibility_info_by_feature[ $feature_id ] = array(
-				'compatible'   => array(),
-				'incompatible' => array(),
+	private function get_feature_definitions() {
+		if ( is_null( $this->features ) ) {
+			$features = array(
+				'analytics'            => array(
+					'name'               => __( 'Analytics', 'woocommerce' ),
+					'description'        => __( 'Enable WooCommerce Analytics', 'woocommerce' ),
+					'option_key'         => Analytics::TOGGLE_OPTION_NAME,
+					'is_experimental'    => false,
+					'enabled_by_default' => true,
+					'disable_ui'         => false,
+					'is_legacy'          => true,
+				),
+				'new_navigation'       => array(
+					'name'            => __( 'Navigation', 'woocommerce' ),
+					'description'     => __( 'Add the new WooCommerce navigation experience to the dashboard', 'woocommerce' ),
+					'option_key'      => Init::TOGGLE_OPTION_NAME,
+					'is_experimental' => false,
+					'disable_ui'      => false,
+					'is_legacy'       => true,
+				),
+				'product_block_editor' => array(
+					'name'            => __( 'New product editor', 'woocommerce' ),
+					'description'     => __( 'Try the new product editor (Beta)', 'woocommerce' ),
+					'option_key'      => NewProductManagementExperience::TOGGLE_OPTION_NAME,
+					'is_experimental' => true,
+					'disable_ui'      => false,
+					'is_legacy'       => true,
+					'disabled'        => function() {
+						return version_compare( get_bloginfo( 'version' ), '6.2', '<' );
+					},
+					'desc_tip'        => function() {
+						$string = '';
+						if ( version_compare( get_bloginfo( 'version' ), '6.2', '<' ) ) {
+							$string = __( '⚠ This feature is compatible with WordPress version 6.2 or higher.', 'woocommerce' );
+						}
+						return $string;
+					},
+				),
+				'cart_checkout_blocks' => array(
+					'name'            => __( 'Cart & Checkout Blocks', 'woocommerce' ),
+					'description'     => __( 'Optimize for faster checkout', 'woocommerce' ),
+					'is_experimental' => false,
+					'disable_ui'      => true,
+				),
+				'marketplace'          => array(
+					'name'               => __( 'Marketplace', 'woocommerce' ),
+					'description'        => __(
+						'New, faster way to find extensions and themes for your WooCommerce store',
+						'woocommerce'
+					),
+					'is_experimental'    => false,
+					'enabled_by_default' => true,
+					'disable_ui'         => false,
+					'is_legacy'          => true,
+				),
 			);
+
+			/**
+			 * Filter the definitions of the features listed on the Features Settings screen.
+			 *
+			 * @since 8.1.0
+			 *
+			 * @param array $features The feature definitions.
+			 */
+			$this->features = apply_filters( 'woocommerce_feature_definitions', $features );
```

Comment:
Do we really want to add this filter? The purpose of the features engine was to have a centralized place to handle WooCommerce _core_ features, without allowing plugins to make any changes, and that's why all the features are hardcoded; but adding this filter defeats the whole thing.

What if instead, the `CustomOrdersTableController->add_feature_definition` method is made public (the class is `Internal` anyway) and we just call it directly?

And I'd say that the same applies to for `woocommerce_feature_setting`.

Reply:
> the purpose of the features engine was to have a centralized place to handle WooCommerce core features, without allowing plugins to make any changes, and that's why all the features are hardcoded; but adding this filter defeats the whole thing.

🤔 Fair point. So instead of a filter, I think you mean we could have `FeaturesController->add_feature_definition` and call that in other classes to add features? It seems like that would work, as long as it didn't create any circular logic loops from classes each depending on each other, which is what happened in my initial attempt. I guess a further question when using the `add_feature_definition` method would be, what action hook should it be called on, to ensure that all the features get registered correctly?

> And I'd say that the same applies to for `woocommerce_feature_setting`

We could potentially still turn this into a method and then deprecate the filter.

I'll work on this and see what I can come up with. Thanks for your review!

Reply:
> you mean we could have FeaturesController->add_feature_definition and call that in other classes to add features?

No, I was thinking of doing it the other way around: other classes would have their own `get_feature_definition` method, and these would be called from inside `FeaturesController::get_feature_definitions`. This would be cleaner I think, since we would be able to see at a glance all the calls to these methods by just looking inside the `FeaturesController` class.

Reply:
Ah, ok. I'm not as keen on that approach for a few reasons:

* Not as much "separation of concerns" since you'd still have to edit the FeaturesController file every time you want to add a new feature, and each feature class would then be a dependency of the FeaturesController
* Related, I'm leery of creating circular dependencies. In the case of `CustomOrdersTableController`, it already has the FeaturesController as an injected dependency for other reasons. If we then try to call a method of it from within the features controller itself, it gets tricky to not have an endless loop situation (as I mentioned elsewhere, I already struggled with this)
* Doesn't provide a standard interface for registering a feature, whereas using `FeaturesController->add_feature_definition` means every feature will register in the same way

> we would be able to see at a glance all the calls to these methods by just looking inside the FeaturesController class.

I understand the benefit of having one place to see which features are being registered, but I'm not sure it outweighs the benefits of keeping the FeaturesController uncoupled from the other parts of the codebase where the features live...

Reply:
Refactored to use a method instead of a filter in 453457ade5a58c2fd94e2a36d66b283660a3e252